buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -352,7 +595,527 @@ static String decodeString(ByteBuffer src, Charset charset) {
         } catch (CharacterCodingException x) {
             throw new IllegalStateException(x);
         }
-        return dst.flip().toString();
+    }
+
+    /**
+     * Returns a multi-line hexadecimal dump of the specified {@link ByteBuf} that is easy to read by humans.
+     */
+    public static String prettyHexDump(ByteBuf buffer) {
+        return prettyHexDump(buffer, buffer.readerIndex(), buffer.readableBytes());
+    }
+
+    /**
+     * Returns a multi-line hexadecimal dump of the specified {@link ByteBuf} that is easy to read by humans,
+     * starting at the given {@code offset} using the given {@code length}.
+     */
+    public static String prettyHexDump(ByteBuf buffer, int offset, int length) {
+        return HexUtil.prettyHexDump(buffer, offset, length);
+    }
+
+    /**
+     * Appends the prettified multi-line hexadecimal dump of the specified {@link ByteBuf} to the specified
+     * {@link StringBuilder} that is easy to read by humans.
+     */
+    public static void appendPrettyHexDump(StringBuilder dump, ByteBuf buf) {
+        appendPrettyHexDump(dump, buf, buf.readerIndex(), buf.readableBytes());
+    }
+
+    /**
+     * Appends the prettified multi-line hexadecimal dump of the specified {@link ByteBuf} to the specified
+     * {@link StringBuilder} that is easy to read by humans, starting at the given {@code offset} using
+     * the given {@code length}.
+     */
+    public static void appendPrettyHexDump(StringBuilder dump, ByteBuf buf, int offset, int length) {
+        HexUtil.appendPrettyHexDump(dump, buf, offset, length);
+    }
+
+    /* Separate class so that the expensive static initialization is only done when needed */
+    private static final class HexUtil {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1118)